File: backend/src/controllers/suivi.controller.js
const DailyHealthLog = require('../models/DailyHealthLog');
const User = require('../models/User');
const { computeHealthScore, summarizeTrend } = require('../utils/healthScore');

function dayStart(d) {
  const x = new Date(d);
  x.setHours(0, 0, 0, 0);
  return x;
}

exports.upsertDaily = async (req, res) => {
  try {
    const date = dayStart(req.body.date || new Date());
    const profile = {
      gender: req.user.gender || 'woman',
      hasDiabetes: Boolean(req.user.hasDiabetes),
    };

    const payload = {
      energy: req.body.energy,
      sleepHours: req.body.sleepHours,
      sleepQuality: req.body.sleepQuality,
      stress: req.body.stress,
      mood: req.body.mood,
      waterLiters: req.body.waterLiters,
      steps: req.body.steps,
      exerciseMinutes: req.body.exerciseMinutes,
      workoutIntensity: req.body.workoutIntensity,
      recovery: req.body.recovery,
      weightKg: req.body.weightKg,
      restingHeartRate: req.body.restingHeartRate,
      fastingGlucose: req.body.fastingGlucose,
      postMealGlucose: req.body.postMealGlucose,
      tookMedication: req.body.tookMedication,
      carbsGrams: req.body.carbsGrams,
      symptoms: req.body.symptoms || [],
      notes: req.body.notes || '',
    };

    const scored = computeHealthScore(payload, profile);

    const log = await DailyHealthLog.findOneAndUpdate(
      { patient: req.user._id, date },
      {
        patient: req.user._id,
        date,
        ...payload,
        healthScore: scored.score,
        healthLabel: scored.label,
      },
      { upsert: true, new: true, setDefaultsOnInsert: true }
    );

    res.status(201).json({
      log,
      prediction: scored,
      message: scored.isGoodHealth
        ? 'Today looks like a good health day based on your inputs.'
        : 'Today needs attention — review tips and consider contacting a clinician if symptoms persist.',
    });
  } catch (error) {
    console.error('upsertDaily:', error);
    res.status(500).json({ message: 'Failed to save daily health log' });
  }
};

exports.getDailyHistory = async (req, res) => {
  try {
    const limit = Math.min(90, Number(req.query.limit) || 30);
    const logs = await DailyHealthLog.find({ patient: req.user._id })
      .sort({ date: -1 })
      .limit(limit);

    const chronological = [...logs].reverse();
    const trend = summarizeTrend(chronological);

    let today = null;
    const todayKey = dayStart(new Date()).toISOString().slice(0, 10);
    const found = logs.find((l) => l.date.toISOString().slice(0, 10) === todayKey);
    if (found) {
      today = {
        log: found,
        prediction: computeHealthScore(found, {
          gender: req.user.gender,
          hasDiabetes: req.user.hasDiabetes,
        }),
      };
    }

    res.json({
      logs,
      trend,
      today,
      chart: chronological.map((l) => ({
        date: l.date.toISOString().slice(0, 10),
        score: l.healthScore,
        label: l.healthLabel,
        fastingGlucose: l.fastingGlucose,
        energy: l.energy,
        stress: l.stress,
      })),
    });
  } catch (error) {
    console.error('getDailyHistory:', error);
    res.status(500).json({ message: 'Failed to load health history' });
  }
};

exports.getToday = async (req, res) => {
  try {
    const date = dayStart(new Date());
    const log = await DailyHealthLog.findOne({ patient: req.user._id, date });
    if (!log) return res.json({ log: null, prediction: null });
    const prediction = computeHealthScore(log, {
      gender: req.user.gender,
      hasDiabetes: req.user.hasDiabetes,
    });
    res.json({ log, prediction });
  } catch (error) {
    res.status(500).json({ message: 'Failed to load today log' });
  }
};

exports.updateDiabetesProfile = async (req, res) => {
  try {
    const { hasDiabetes, diabetesType } = req.body;
    if (hasDiabetes !== undefined) req.user.hasDiabetes = Boolean(hasDiabetes);
    if (diabetesType !== undefined) req.user.diabetesType = diabetesType;
    await req.user.save();
    res.json({ user: req.user.toSafeJSON() });
  } catch (error) {
    res.status(500).json({ message: 'Failed to update diabetes profile' });
  }
};

exports.getDiabetesOverview = async (req, res) => {
  try {
    const logs = await DailyHealthLog.find({ patient: req.user._id })
      .sort({ date: -1 })
      .limit(30);

    const withGlucose = logs.filter(
      (l) => l.fastingGlucose != null || l.postMealGlucose != null
    );
    const fasting = withGlucose
      .map((l) => l.fastingGlucose)
      .filter((v) => v != null);
    const avgFasting = fasting.length
      ? Math.round(fasting.reduce((a, b) => a + b, 0) / fasting.length)
      : null;

    const medDays = logs.filter((l) => l.tookMedication).length;
    const adherence =
      logs.length > 0 ? Math.round((medDays / logs.length) * 100) : null;

    res.json({
      profile: {
        hasDiabetes: req.user.hasDiabetes,
        diabetesType: req.user.diabetesType,
      },
      avgFasting,
      adherencePercent: adherence,
      recent: withGlucose.slice(0, 14).map((l) => ({
        date: l.date.toISOString().slice(0, 10),
        fastingGlucose: l.fastingGlucose,
        postMealGlucose: l.postMealGlucose,
        tookMedication: l.tookMedication,
        carbsGrams: l.carbsGrams,
        healthScore: l.healthScore,
      })),
      chart: [...logs]
        .reverse()
        .filter((l) => l.fastingGlucose != null)
        .map((l) => ({
          date: l.date.toISOString().slice(0, 10),
          fastingGlucose: l.fastingGlucose,
          postMealGlucose: l.postMealGlucose,
        })),
    });
  } catch (error) {
    console.error('getDiabetesOverview:', error);
    res.status(500).json({ message: 'Failed to load diabetes overview' });
  }
};

File: backend-dotnet/MyHeath.Api/Controllers/SuiviController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using MongoDB.Driver;
using MyHeath.Api.Infrastructure;
using MyHeath.Api.Models;
using MyHeath.Api.Services;

namespace MyHeath.Api.Controllers;

[ApiController]
[Route("api/suivi")]
[Authorize(Roles = Roles.Patient)]
public class SuiviController : ApiControllerBase
{
    public SuiviController(MongoContext db) : base(db) { }

    public record DailyDto(
        DateTime? Date, int? Energy, double? SleepHours, int? SleepQuality, int? Stress,
        string? Mood, double? WaterLiters, int? Steps, int? ExerciseMinutes,
        string? WorkoutIntensity, int? Recovery, double? WeightKg, int? RestingHeartRate,
        double? FastingGlucose, double? PostMealGlucose, bool? TookMedication,
        double? CarbsGrams, List<string>? Symptoms, string? Notes);

    public record DiabetesProfileDto(bool? HasDiabetes, string? DiabetesType);

    [HttpPost("daily")]
    public async Task<IActionResult> UpsertDaily([FromBody] DailyDto dto)
    {
        var user = await CurrentUserAsync();
        if (user is null) return Fail("Unauthorized", 401);
        var date = (dto.Date ?? DateTime.UtcNow).Date;
        var existing = await Db.DailyHealthLogs.Find(l => l.Patient == user.Id && l.Date == date).FirstOrDefaultAsync();
        var log = existing ?? new DailyHealthLog { Patient = user.Id, Date = date };
        Apply(log, dto);
        var (score, label) = HealthScoreService.Compute(log, user);
        log.HealthScore = score;
        log.HealthLabel = label;
        log.UpdatedAt = DateTime.UtcNow;
        if (existing is null) await Db.DailyHealthLogs.InsertOneAsync(log);
        else await Db.DailyHealthLogs.ReplaceOneAsync(l => l.Id == log.Id, log);

        return StatusCode(201, new
        {
            log,
            prediction = new { score, label, message = $"Predicted wellness: {label} ({score}/100)" },
            message = "Daily health saved"
        });
    }

    [HttpGet("daily")]
    public async Task<IActionResult> History([FromQuery] int limit = 30)
    {
        var user = await CurrentUserAsync();
        if (user is null) return Fail("Unauthorized", 401);
        limit = Math.Clamp(limit, 1, 90);
        var logs = await Db.DailyHealthLogs.Find(l => l.Patient == user.Id)
            .SortByDescending(l => l.Date).Limit(limit).ToListAsync();
        var today = DateTime.UtcNow.Date;
        var todayLog = logs.FirstOrDefault(l => l.Date.Date == today);
        var chart = logs.OrderBy(l => l.Date).Select(l => new
        {
            date = l.Date.ToString("yyyy-MM-dd"),
            score = l.HealthScore,
            fasting = l.FastingGlucose
        });
        var avg = logs.Count > 0 ? (int)Math.Round(logs.Average(l => l.HealthScore)) : 0;
        return Ok(new
        {
            logs,
            trend = new { averageScore = avg, count = logs.Count },
            today = todayLog,
            chart
        });
    }

    [HttpGet("daily/today")]
    public async Task<IActionResult> Today()
    {
        var user = await CurrentUserAsync();
        if (user is null) return Fail("Unauthorized", 401);
        var today = DateTime.UtcNow.Date;
        var log = await Db.DailyHealthLogs.Find(l => l.Patient == user.Id && l.Date == today).FirstOrDefaultAsync();
        object? prediction = log is null ? null : new { score = log.HealthScore, label = log.HealthLabel };
        return Ok(new { log, prediction });
    }

    [HttpGet("diabetes")]
    public async Task<IActionResult> Diabetes()
    {
        var user = await CurrentUserAsync();
        if (user is null) return Fail("Unauthorized", 401);
        var recent = await Db.DailyHealthLogs.Find(l => l.Patient == user.Id)
            .SortByDescending(l => l.Date).Limit(30).ToListAsync();
        var withFasting = recent.Where(l => l.FastingGlucose != null).ToList();
        var avgFasting = withFasting.Count > 0 ? withFasting.Average(l => l.FastingGlucose!.Value) : (double?)null;
        var medDays = recent.Count(l => l.TookMedication);
        var adherence = recent.Count > 0 ? (int)Math.Round(100.0 * medDays / recent.Count) : 0;
        return Ok(new
        {
            profile = new { user.HasDiabetes, user.DiabetesType },
            avgFasting,
            adherencePercent = adherence,
            recent,
            chart = recent.OrderBy(l => l.Date).Select(l => new
            {
                date = l.Date.ToString("yyyy-MM-dd"),
                fasting = l.FastingGlucose,
                postMeal = l.PostMealGlucose
            })
        });
    }

    [HttpPut("diabetes/profile")]
    public async Task<IActionResult> UpdateDiabetes([FromBody] DiabetesProfileDto dto)
    {
        var user = await CurrentUserAsync();
        if (user is null) return Fail("Unauthorized", 401);
        if (dto.HasDiabetes.HasValue) user.HasDiabetes = dto.HasDiabetes.Value;
        if (dto.DiabetesType != null) user.DiabetesType = dto.DiabetesType;
        user.UpdatedAt = DateTime.UtcNow;
        await Db.Users.ReplaceOneAsync(u => u.Id == user.Id, user);
        return Ok(new { user = UserMapper.ToSafe(user) });
    }

    private static void Apply(DailyHealthLog log, DailyDto dto)
    {
        if (dto.Energy.HasValue) log.Energy = dto.Energy.Value;
        if (dto.SleepHours.HasValue) log.SleepHours = dto.SleepHours.Value;
        if (dto.SleepQuality.HasValue) log.SleepQuality = dto.SleepQuality.Value;
        if (dto.Stress.HasValue) log.Stress = dto.Stress.Value;
        if (dto.Mood != null) log.Mood = dto.Mood;
        if (dto.WaterLiters.HasValue) log.WaterLiters = dto.WaterLiters.Value;
        if (dto.Steps.HasValue) log.Steps = dto.Steps.Value;
        if (dto.ExerciseMinutes.HasValue) log.ExerciseMinutes = dto.ExerciseMinutes.Value;
        if (dto.WorkoutIntensity != null) log.WorkoutIntensity = dto.WorkoutIntensity;
        if (dto.Recovery.HasValue) log.Recovery = dto.Recovery.Value;
        if (dto.WeightKg.HasValue) log.WeightKg = dto.WeightKg;
        if (dto.RestingHeartRate.HasValue) log.RestingHeartRate = dto.RestingHeartRate;
        if (dto.FastingGlucose.HasValue) log.FastingGlucose = dto.FastingGlucose;
        if (dto.PostMealGlucose.HasValue) log.PostMealGlucose = dto.PostMealGlucose;
        if (dto.TookMedication.HasValue) log.TookMedication = dto.TookMedication.Value;
        if (dto.CarbsGrams.HasValue) log.CarbsGrams = dto.CarbsGrams;
        if (dto.Symptoms != null) log.Symptoms = dto.Symptoms;
        if (dto.Notes != null) log.Notes = dto.Notes;
    }
}
